### ISSUE TITLE
Support merging tool metadata from additional (uncommitted) files

### ISSUE THREAD

**Issue by maciektr** [software-mansion/scarb#677]
### Problem

Some tools may require config keys that should not be written into the manifest file. Some of those keys may be switched based on profile / other user input (example: profile credentials in aws cli). This config keys should co-exist with currently used `tool` section based config design. 

### Proposed Solution

Research solutions used by established tool ecosystems. 

### Notes

cc @MaksymilianDemitraszek @cptartur

**Comment by MaksymilianDemitraszek** [comment]
Contex: providing private configuration in the toml
```
[tool.forge]
[A]
address = $address_env1
block_depth = 12312

[B]
address = $address_env2
block_depth = 12
```

**Comment by mkaput** [comment]
FYI The only tool that has this, that I'm aware of, is Terraform.

**Comment by MaksymilianDemitraszek** [comment]
@mkaput we are effectively trying to build starknet terraform :D

**Comment by cptartur** [comment]
> OK, so being inspired by https://doc.rust-lang.org/cargo/reference/config.html and https://developer.hashicorp.com/terraform/language/values/variables#variable-definitions-tfvars-files I propose following solution:
> 
> We will allow reading `[tool]` section values from other files, namely:
> 
> * `config.toml`
> * `[prefix].config.toml`, i.e. any file which name ends with `.config.toml`
> 
> Let's call these **configuration files**. This system will enable local configuration for a particular package as well as global configuration. Scarb will look for configuration files in:
> 
> 1. `[package root]/.scarb`
> 2. `[workspace root]/.scarb`
> 3. [`$SCARB_CONFIG`](https://docs.swmansion.com/scarb/docs/reference/global-directories.html#config-directory)
> 
> The order of loading will be:
> 
> 1. `$SCARB_CONFIG/config.toml`
> 2. `$SCARB_CONFIG/*.config.toml`, sorted lexicographically
> 3. `[workspace root]/Scarb.toml` & `[package root]/Scarb.toml`, overridden via existing `.workspace = true` mechanism
> 4. `[workspace root]/.scarb/config.toml`
> 5. `[workspace root]/.scarb/*.config.toml`, sorted lexicographically
> 6. `[package root]/.scarb/config.toml`
> 7. `[package root]/.scarb/*.config.toml`, sorted lexicographically
> 

Shouldn't package roots be resolved before the workspace root? Though I might just be confused.

> The configuration files will have a subset of Scarb.toml syntax, i.e. they'll have to have the `[tool]` section. For example:
> 
> ```toml
> [tool.forge]
> [A]
> address = "some address"
> ```
> 
> Merging will be applied top-down, i.e. later entries will override earlier ones.
> 
> **Problem:** I don't think we can use the existing [`toml_merge` algorithm](https://github.com/software-mansion/scarb/blob/c412b06b0e1069ea1ca503c945d5e8dfd12bb398/scarb/src/internal/serdex.rs#L8-L25), here. I am also not in favour of adding a new deep-merging algorithm. Perhaps we could have `.config = "config var name"` mechanism, similar to `.workspace` one?

What do you mean exactly by this mechanism? That you will be required to define something like `my_arg.config = true` if you want it to be "inherited" from the necessary file?

**Comment by maciektr** [comment]
> What do you mean exactly by this mechanism? That you will be required to define something like my_arg.config = true if you want it to be "inherited" from the necessary file?
@cptartur 
Yes, analogous to inheriting keys from workspace manifest with `key.workspace = true` we could have some mechanism to inherit from config files.
